diff --git a/docs/settings.rst b/docs/settings.rst
index 72c72fb1e..8e634d8f9 100644
--- a/docs/settings.rst
+++ b/docs/settings.rst
@@ -150,13 +150,17 @@ Basic settings
 
       READERS = {'foo': FooReader}
 
-.. data:: IGNORE_FILES = ['.*']
+.. data:: IGNORE_FILES = ['**/.*']
 
-   A list of glob patterns. Files and directories matching any of these patterns
-   will be ignored by the processor. For example, the default ``['.*']`` will
+   A list of Unix glob patterns. Files and directories matching any of these patterns
+   or any of the commonly hidden files and directories set by ``watchfiles.DefaultFilter``
+   will be ignored by the processor. For example, the default ``['**/.*']`` will
    ignore "hidden" files and directories, and ``['__pycache__']`` would ignore
    Python 3's bytecode caches.
 
+   For a full list of the commonly hidden files set by ``watchfiles.DefaultFilter``,
+   please refer to the `watchfiles documentation`_.
+
 .. data:: MARKDOWN = {...}
 
    Extra configuration settings for the Markdown processor. Refer to the Python
@@ -1423,3 +1427,4 @@ Example settings
 
 .. _Jinja Environment documentation: https://jinja.palletsprojects.com/en/latest/api/#jinja2.Environment
 .. _Docutils Configuration: http://docutils.sourceforge.net/docs/user/config.html
+.. _`watchfiles documentation`: https://watchfiles.helpmanual.io/api/filters/#watchfiles.DefaultFilter.ignore_dirs
diff --git a/pelican/settings.py b/pelican/settings.py
index 1fb0b2111..3abe11bbe 100644
--- a/pelican/settings.py
+++ b/pelican/settings.py
@@ -158,7 +158,7 @@ def load_source(name: str, path: str) -> ModuleType:
     "PYGMENTS_RST_OPTIONS": {},
     "TEMPLATE_PAGES": {},
     "TEMPLATE_EXTENSIONS": [".html"],
-    "IGNORE_FILES": [".*"],
+    "IGNORE_FILES": ["**/.*"],
     "SLUG_REGEX_SUBSTITUTIONS": [
         (r"[^\w\s-]", ""),  # remove non-alphabetical/whitespace/'-' chars
         (r"(?u)\A\s*", ""),  # strip leading whitespace
diff --git a/pelican/utils.py b/pelican/utils.py
index 21dbe804e..75889a7bb 100644
--- a/pelican/utils.py
+++ b/pelican/utils.py
@@ -811,15 +811,30 @@ def order_content(
     return content_list
 
 
+class FileChangeFilter(watchfiles.DefaultFilter):
+    def __init__(self, ignore_file_patterns: Sequence[str], *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.ignore_file_patterns = ignore_file_patterns
+
+    def __call__(self, change: watchfiles.Change, path: str) -> bool:
+        """Returns `True` if a file should be watched for changes. The `IGNORE_FILES`
+        setting is a list of Unix glob patterns. This call will filter out files and
+        directories specified by `IGNORE_FILES` Pelican setting and by the default
+        filters of `watchfiles.DefaultFilter`, seen here:
+        https://watchfiles.helpmanual.io/api/filters/#watchfiles.DefaultFilter.ignore_dirs
+        """
+        return super().__call__(change, path) and not any(
+            fnmatch.fnmatch(os.path.abspath(path), p) for p in self.ignore_file_patterns
+        )
+
+
 def wait_for_changes(
     settings_file: str,
     settings: Settings,
 ) -> set[tuple[Change, str]]:
     content_path = settings.get("PATH", "")
     theme_path = settings.get("THEME", "")
-    ignore_files = {
-        fnmatch.translate(pattern) for pattern in settings.get("IGNORE_FILES", [])
-    }
+    ignore_file_patterns = set(settings.get("IGNORE_FILES", []))
 
     candidate_paths = [
         settings_file,
@@ -844,7 +859,7 @@ def wait_for_changes(
     return next(
         watchfiles.watch(
             *watching_paths,
-            watch_filter=watchfiles.DefaultFilter(ignore_entity_patterns=ignore_files),  # type: ignore
+            watch_filter=FileChangeFilter(ignore_file_patterns=ignore_file_patterns),
             rust_timeout=0,
         )
     )
