diff --git a/pelican/utils.py b/pelican/utils.py
index f3a012176..de6ef9bfe 100644
--- a/pelican/utils.py
+++ b/pelican/utils.py
@@ -155,7 +155,9 @@ def __repr__(self):
 
     def __get__(self, obj, objtype):
         '''Support instance methods.'''
-        return partial(self.__call__, obj)
+        fn = partial(self.__call__, obj)
+        fn.cache = self.cache
+        return fn
 
 
 def deprecated_attribute(old, new, since=None, remove=None, doc=None):
