diff --git a/pytype/tools/analyze_project/pytype_runner.py b/pytype/tools/analyze_project/pytype_runner.py
index 6812860f0..1c12b6e22 100644
--- a/pytype/tools/analyze_project/pytype_runner.py
+++ b/pytype/tools/analyze_project/pytype_runner.py
@@ -196,9 +196,6 @@ def get_module_action(self, module):
     else:
       action = Action.INFER
       report = logging.info
-    if not f.endswith('.py'):
-      report('skipped: non-Python file %s', f)
-      return None
     # For builtin and system files, do not attempt to generate a pyi.
     if module.kind in ('Builtin', 'System'):
       action = Action.GENERATE_DEFAULT
