diff --git a/pelican/tests/content/article_with_attributes_containing_double_quotes.html b/pelican/tests/content/article_with_attributes_containing_double_quotes.html
new file mode 100644
index 000000000..7daa58015
--- /dev/null
+++ b/pelican/tests/content/article_with_attributes_containing_double_quotes.html
@@ -0,0 +1,11 @@
+<html>
+    <head>
+    </head>
+    <body>
+        Ensure that if an attribute value contains a double quote, it is
+        surrounded with single quotes, otherwise with double quotes.
+        <span data-test="'single quoted string'">Span content</span>
+        <span data-test='"double quoted string"'>Span content</span>
+        <span data-test="string without quotes">Span content</span>
+    </body>
+</html>
diff --git a/pelican/tests/test_cache.py b/pelican/tests/test_cache.py
index 006e421b6..3da3f7897 100644
--- a/pelican/tests/test_cache.py
+++ b/pelican/tests/test_cache.py
@@ -61,7 +61,7 @@ def test_article_object_caching(self):
         - article_with_null_attributes.html
         - 2012-11-30_md_w_filename_meta#foo-bar.md
         """
-        self.assertEqual(generator.readers.read_file.call_count, 3)
+        self.assertEqual(generator.readers.read_file.call_count, 4)
 
     @unittest.skipUnless(MagicMock, 'Needs Mock module')
     def test_article_reader_content_caching(self):
diff --git a/pelican/tests/test_readers.py b/pelican/tests/test_readers.py
index 5fabc470c..dc4348357 100644
--- a/pelican/tests/test_readers.py
+++ b/pelican/tests/test_readers.py
@@ -587,6 +587,17 @@ def test_article_with_null_attributes(self):
         <input name="test" disabled style="" />
     ''', page.content)
 
+    def test_article_with_attributes_containing_double_quotes(self):
+        page = self.read_file(path='article_with_attributes_containing_' +
+                                   'double_quotes.html')
+        self.assertEqual('''
+        Ensure that if an attribute value contains a double quote, it is
+        surrounded with single quotes, otherwise with double quotes.
+        <span data-test="'single quoted string'">Span content</span>
+        <span data-test='"double quoted string"'>Span content</span>
+        <span data-test="string without quotes">Span content</span>
+    ''', page.content)
+
     def test_article_metadata_key_lowercase(self):
         # Keys of metadata should be lowercase.
         page = self.read_file(path='article_with_uppercase_metadata.html')
