diff --git a/doc/changes.rst b/doc/changes.rst
index 4f4f91afb2..adaeafc9da 100644
--- a/doc/changes.rst
+++ b/doc/changes.rst
@@ -4,6 +4,29 @@ Change log
 Stable versions
 ~~~~~~~~~~~~~~~
 
+Version 2.2.0 ()
+-----------------------------------
+
+Breaking Changes
+^^^^^^^^^^^^^^^^
+
+* The ``github.Comparison.Comparison`` instance returned by ``Repository.compare`` provides a ``commits``
+  property that used to return a ``list[github.Commit.Commit]``, which has now been changed
+  to ``PaginatedList[github.Commit.Commit]``. This breaks user code that assumes a ``list``:
+
+.. code-block:: python
+
+    commits = repo.compare("v0.6", "v0.7").commits
+    no_of_commits = len(commits)
+
+This will raise a ``TypeError: object of type 'PaginatedList' has no len()``, as the returned ``PaginatedList``
+does not support the ``len()`` method. Use the ``totalCount`` property instead:
+
+.. code-block:: python
+
+    commits = repo.compare("v0.6", "v0.7").commits
+    no_of_commits = commits.totalCount
+
 Version 2.1.1 (September 29, 2023)
 -----------------------------------
 
diff --git a/github/Comparison.py b/github/Comparison.py
index a147d3c555..d2930fcac8 100644
--- a/github/Comparison.py
+++ b/github/Comparison.py
@@ -40,6 +40,7 @@
 import github.Commit
 import github.File
 from github.GithubObject import Attribute, CompletableGithubObject, NotSet
+from github.PaginatedList import PaginatedList
 
 
 class Comparison(CompletableGithubObject):
@@ -51,7 +52,6 @@ def _initAttributes(self) -> None:
         self._ahead_by: Attribute[int] = NotSet
         self._base_commit: Attribute[github.Commit.Commit] = NotSet
         self._behind_by: Attribute[int] = NotSet
-        self._commits: Attribute[list[github.Commit.Commit]] = NotSet
         self._diff_url: Attribute[str] = NotSet
         self._files: Attribute[list[github.File.File]] = NotSet
         self._html_url: Attribute[str] = NotSet
@@ -80,10 +80,21 @@ def behind_by(self) -> int:
         self._completeIfNotSet(self._behind_by)
         return self._behind_by.value
 
+    # This should be a method, but this used to be a property and cannot be changed without breaking user code
+    # TODO: remove @property on version 3
     @property
-    def commits(self) -> list[github.Commit.Commit]:
-        self._completeIfNotSet(self._commits)
-        return self._commits.value
+    def commits(self) -> PaginatedList[github.Commit.Commit]:
+        return PaginatedList(
+            github.Commit.Commit,
+            self._requester,
+            self.url,
+            {},
+            None,
+            "commits",
+            "total_commits",
+            self.raw_data,
+            self.raw_headers,
+        )
 
     @property
     def diff_url(self) -> str:
@@ -137,8 +148,6 @@ def _useAttributes(self, attributes: dict[str, Any]) -> None:
             self._base_commit = self._makeClassAttribute(github.Commit.Commit, attributes["base_commit"])
         if "behind_by" in attributes:  # pragma no branch
             self._behind_by = self._makeIntAttribute(attributes["behind_by"])
-        if "commits" in attributes:  # pragma no branch
-            self._commits = self._makeListOfClassesAttribute(github.Commit.Commit, attributes["commits"])
         if "diff_url" in attributes:  # pragma no branch
             self._diff_url = self._makeStringAttribute(attributes["diff_url"])
         if "files" in attributes:  # pragma no branch
diff --git a/github/EnterpriseConsumedLicenses.py b/github/EnterpriseConsumedLicenses.py
index 1cae2700ba..d05e47b6da 100644
--- a/github/EnterpriseConsumedLicenses.py
+++ b/github/EnterpriseConsumedLicenses.py
@@ -87,8 +87,8 @@ def get_users(self) -> PaginatedList[NamedEnterpriseUser]:
             url_parameters,
             None,
             "users",
-            self.raw_data,
-            self.raw_headers,
+            firstData=self.raw_data,
+            firstHeaders=self.raw_headers,
         )
 
     def _useAttributes(self, attributes: Dict[str, Any]) -> None:
diff --git a/github/PaginatedList.py b/github/PaginatedList.py
index 29cdb42d9e..719797c661 100644
--- a/github/PaginatedList.py
+++ b/github/PaginatedList.py
@@ -152,6 +152,7 @@ def __init__(
         firstParams: Any,
         headers: Optional[Dict[str, str]] = None,
         list_item: str = "items",
+        total_count_item: str = "total_count",
         firstData: Optional[Any] = None,
         firstHeaders: Optional[Dict[str, Union[str, int]]] = None,
         attributesTransformer: Optional[Callable[[Dict[str, Any]], Dict[str, Any]]] = None,
@@ -164,6 +165,7 @@ def __init__(
         self.__nextParams = firstParams or {}
         self.__headers = headers
         self.__list_item = list_item
+        self.__total_count_item = total_count_item
         if self.__requester.per_page != 30:
             self.__nextParams["per_page"] = self.__requester.per_page
         self._reversed = False
@@ -255,7 +257,7 @@ def _getPage(self, data: Any, headers: Dict[str, Any]) -> List[T]:
                 self.__nextUrl = links["next"]
         self.__nextParams = None
         if self.__list_item in data:
-            self.__totalCount = data.get("total_count")
+            self.__totalCount = data.get(self.__total_count_item)
             data = data[self.__list_item]
         content = [
             self.__contentClass(self.__requester, headers, self._transformAttributes(element), completed=False)
diff --git a/github/Repository.py b/github/Repository.py
index 336a31e49a..27ea817e4d 100644
--- a/github/Repository.py
+++ b/github/Repository.py
@@ -1105,7 +1105,7 @@ def remove_invitation(self, invite_id: int) -> None:
 
     def compare(self, base: str, head: str) -> Comparison:
         """
-        :calls: `GET /repos/{owner}/{repo}/compare/{base...:head} <https://docs.github.com/en/rest/reference/repos#commits>`_
+        :calls: `GET /repos/{owner}/{repo}/compare/{base...:head} <https://docs.github.com/en/rest/commits/commits#compare-two-commits>`_
         :param base: string
         :param head: string
         :rtype: :class:`github.Comparison.Comparison`
@@ -1114,7 +1114,11 @@ def compare(self, base: str, head: str) -> Comparison:
         assert isinstance(head, str), head
         base = urllib.parse.quote(base)
         head = urllib.parse.quote(head)
-        headers, data = self._requester.requestJsonAndCheck("GET", f"{self.url}/compare/{base}...{head}")
+        # the compare API has a per_page default of 250, which is different to Consts.DEFAULT_PER_PAGE
+        per_page = self._requester.per_page if self._requester.per_page != Consts.DEFAULT_PER_PAGE else 250
+        # only with page=1 we get the pagination headers for the commits element
+        params = {"page": 1, "per_page": per_page}
+        headers, data = self._requester.requestJsonAndCheck("GET", f"{self.url}/compare/{base}...{head}", params)
         return github.Comparison.Comparison(self._requester, headers, data, completed=True)
 
     def create_autolink(
